Race condition  in TestChannelNamespaceDefaulting.

**Describe the bug**

There appears to be a race condition between updating the configmap and webhook defaulting getting set up properly.


https://github.com/knative/eventing/runs/1381792073?check_suite_focus=true

```

Failed to get test-default-broker-with-many-triggers-test-default-brokernckf8/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokerhhlmf/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokerp9k4n/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokervmbgq/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokernckf8/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokerhhlmf/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokerp9k4n/default: brokers.eventing.knative.dev "default" not foundFailed to get test-default-broker-with-many-triggers-test-default-brokervmbgq/default: brokers.eventing.knative.dev "default" not foundFailed to get test-trigger-dependency-annotation-d582m/default: brokers.eventing.knative.dev "default" not foundFailed to get test-ping-source-v1-beta1-event-types-4nlqd/default: brokers.eventing.knative.dev "default" not foundFailed to get test-ping-source-v1-alpha2-event-types-c8kxs/default: brokers.eventing.knative.dev "default" not foundFailed to get test-api-server-source-v1-beta1-event-types-cbhwq/default: brokers.eventing.knative.dev "default" not foundFailed to get test-api-server-source-v1-alpha2-event-types-4xf8s/default: brokers.eventing.knative.dev "default" not foundFailed to get test-api-server-source-v1-event-types-bj2vr/default: brokers.eventing.knative.dev "default" not found--- FAIL: TestChannelNamespaceDefaulting (1.28s)
    test_runner.go:153: namespace is : "test-channel-namespace-defaulting-sn9wq"
    namespace_default_channel_test.go:54: Updating defaulting ConfigMap
    namespace_default_channel_test.go:91: CM updated - new values: |
          clusterDefault:
            apiVersion: messaging.knative.dev/v1
            kind: InMemoryChannel
          namespaceDefaults:
            test-channel-namespace-defaulting-sn9wq:
              apiVersion: messaging.knative.dev/v1
              kind: InMemoryChannel
              spec:
                delivery:
                  backoffDelay: PT0.5S
                  backoffPolicy: exponential
                  retry: 5
        
    k8s_events.go:66: EventListener stopped, 0 events seen
    test_runner.go:223: Export logs in "test-channel-namespace-defaulting-sn9wq" to "/home/runner/work/eventing/eventing/artifacts/pod-logs"
panic: interface conversion: interface {} is nil, not map[string]interface {} [recovered]
	panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 2221 [running]:
testing.tRunner.func1.1(0x2928ec0, 0xc000fb0300)
	/opt/hostedtoolcache/go/1.15.3/x64/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc000947c80)
	/opt/hostedtoolcache/go/1.15.3/x64/src/testing/testing.go:1075 +0x636
panic(0x2928ec0, 0xc000fb0300)
	/opt/hostedtoolcache/go/1.15.3/x64/src/runtime/panic.go:975 +0x47a
knative.dev/eventing/test/e2e.TestChannelNamespaceDefaulting(0xc000947c80)
	/home/runner/work/eventing/eventing/src/knative.dev/eventing/test/e2e/namespace_default_channel_test.go:118 +0xc95
testing.tRunner(0xc000947c80, 0x2c83440)
	/opt/hostedtoolcache/go/1.15.3/x64/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.15.3/x64/src/testing/testing.go:1168 +0x5bc
```


A clear and concise description of what the bug is.

 **Expected behavior**
A clear and concise description of what you expected to happen.

 **To Reproduce**
Steps to reproduce the behavior.

 **Knative release version**

 **Additional context**
Add any other context about the problem here such as proposed priority

--- a/test/e2e/namespace_default_channel_test.go
+++ b/test/e2e/namespace_default_channel_test.go
@@ -20,6 +20,8 @@ package e2e
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"testing"
 	"time"
 
@@ -28,12 +30,14 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/util/retry"
 	"knative.dev/pkg/system"
 
 	eventingduck "knative.dev/eventing/pkg/apis/duck/v1"
+	messagingv1 "knative.dev/eventing/pkg/apis/messaging/v1"
 	testlib "knative.dev/eventing/test/lib"
 )
 
@@ -56,20 +60,19 @@ func TestChannelNamespaceDefaulting(t *testing.T) {
 		cm, err := c.Kube.CoreV1().ConfigMaps(system.Namespace()).Get(ctx, defaultChannelCM, metav1.GetOptions{})
 		assert.Nil(t, err)
 
-		defaults := make(map[string]interface{})
-		err = yaml.Unmarshal([]byte(cm.Data["default-ch-config"]), defaults)
+		// Preserve existing namespace defaults.
+		defaults := make(map[string]map[string]interface{})
+		err = yaml.Unmarshal([]byte(cm.Data[defaultChannelConfigKey]), defaults)
 		assert.Nil(t, err)
 
-		defaults["namespaceDefaults"] = map[string]interface{}{
-			c.Namespace: map[string]interface{}{
-				"apiVersion": "messaging.knative.dev/v1",
-				"kind":       "InMemoryChannel",
-				"spec": map[string]interface{}{
-					"delivery": map[string]interface{}{
-						"retry":         5,
-						"backoffPolicy": "exponential",
-						"backoffDelay":  "PT0.5S",
-					},
+		defaults["namespaceDefaults"][c.Namespace] = map[string]interface{}{
+			"apiVersion": "messaging.knative.dev/v1",
+			"kind":       "InMemoryChannel",
+			"spec": map[string]interface{}{
+				"delivery": map[string]interface{}{
+					"retry":         5,
+					"backoffPolicy": "exponential",
+					"backoffDelay":  "PT0.5S",
 				},
 			},
 		}
@@ -95,35 +98,69 @@ func TestChannelNamespaceDefaulting(t *testing.T) {
 	})
 	assert.Nil(t, err)
 
-	obj := &unstructured.Unstructured{
-		Object: map[string]interface{}{
-			"apiVersion": "messaging.knative.dev/v1",
-			"kind":       "Channel",
-			"metadata": map[string]interface{}{
-				"name":      "xyz",
-				"namespace": c.Namespace,
-			},
-		},
+	// Create a Channel and check whether it has the DeliverySpec set as we specified above.
+	// Since the webhook receives the updates at some undetermined time after the update to reduce flakiness retry after
+	// a delay and check whether it has the shape we want it to have.
+
+	namePrefix := "xyz"
+	n := 0
+	lastName := ""
+
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   1.0,
+		Jitter:   0.1,
+		Steps:    5,
 	}
 
-	createdObj, err := c.Dynamic.
-		Resource(schema.GroupVersionResource{Group: "messaging.knative.dev", Version: "v1", Resource: "channels"}).
-		Namespace(c.Namespace).
-		Create(ctx, obj, metav1.CreateOptions{})
+	err = retry.OnError(backoff, func(err error) bool { return err != nil }, func() error {
+		n++
+		name := fmt.Sprintf("%s-%d", namePrefix, n)
+		lastName = name
 
+		obj := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "messaging.knative.dev/v1",
+				"kind":       "Channel",
+				"metadata": map[string]interface{}{
+					"name":      name,
+					"namespace": c.Namespace,
+				},
+			},
+		}
+
+		createdObj, err := c.Dynamic.
+			Resource(schema.GroupVersionResource{Group: "messaging.knative.dev", Version: "v1", Resource: "channels"}).
+			Namespace(c.Namespace).
+			Create(ctx, obj, metav1.CreateOptions{})
+		assert.Nil(t, err)
+
+		channel := &messagingv1.Channel{}
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(createdObj.Object, channel)
+		assert.Nil(t, err)
+
+		if !webhookObservedUpdate(channel) {
+			return fmt.Errorf("webhook hasn't seen the update: %+v", channel)
+		}
+
+		delivery := struct {
+			Delivery eventingduck.DeliverySpec `json:"delivery"`
+		}{}
+		err = json.Unmarshal(channel.Spec.ChannelTemplate.Spec.Raw, &delivery)
+		if err != nil || !webhookObservedUpdateFromDeliverySpec(&delivery.Delivery) {
+			return fmt.Errorf("webhook hasn't seen the update: %+v %+v", delivery, string(channel.Spec.ChannelTemplate.Spec.Raw))
+		}
+
+		assert.Equal(t, "PT0.5S", *delivery.Delivery.BackoffDelay)
+		assert.Equal(t, int32(5), *delivery.Delivery.Retry)
+		assert.Equal(t, eventingduck.BackoffPolicyExponential, *delivery.Delivery.BackoffPolicy)
+
+		return nil
+	})
 	assert.Nil(t, err)
 
-	spec := createdObj.Object["spec"].(map[string]interface{})
-	spec = spec["channelTemplate"].(map[string]interface{})
-	spec = spec["spec"].(map[string]interface{})
-	spec = spec["delivery"].(map[string]interface{})
-
-	assert.Equal(t, "PT0.5S", spec["backoffDelay"])
-	assert.Equal(t, int64(5), spec["retry"])
-	assert.Equal(t, "exponential", spec["backoffPolicy"])
-
-	wait.Poll(time.Second, time.Minute, func() (done bool, err error) {
-		imc, err := c.Eventing.MessagingV1().InMemoryChannels(c.Namespace).Get(ctx, "xyz", metav1.GetOptions{})
+	err = wait.Poll(time.Second, time.Minute, func() (done bool, err error) {
+		imc, err := c.Eventing.MessagingV1().InMemoryChannels(c.Namespace).Get(ctx, lastName, metav1.GetOptions{})
 		if apierrors.IsNotFound(err) {
 			return false, nil
 		}
@@ -135,4 +172,14 @@ func TestChannelNamespaceDefaulting(t *testing.T) {
 
 		return true, nil
 	})
+	assert.Nil(t, err)
+}
+
+func webhookObservedUpdate(ch *messagingv1.Channel) bool {
+	return ch.Spec.ChannelTemplate != nil &&
+		ch.Spec.ChannelTemplate.Spec != nil
+}
+
+func webhookObservedUpdateFromDeliverySpec(d *eventingduck.DeliverySpec) bool {
+	return d.BackoffDelay != nil && d.Retry != nil && d.BackoffPolicy != nil
 }

